### DISCUSSION TITLE
fetcher.data TS type for a POST Form

### DISCUSSION THREAD

**Discussion by wild-lotus** [remix-run/remix#7712] [Q&A]
I am having trouble getting the right type for a `fetcher.data` mutation data after v2 update. I searched a bit but I did not find anything on it. Here is a simple example:

``` ts
export async function action({ request }: ActionFunctionArgs) {
  // ...
  if (error) {
    return json({ error });
  }
  return null;
}

export default function Route() {
  const fetcher = useFetcher();
  return (
    <fetcher.Form method="POST">
      {/* ... */}
      {fetcher.data?.error && <div>{fetcher.data?.error}</div>}
    </fetcher.Form>
  );
}
```
`fetcher.data?.error` used to be ok, but now `fetcher.data` is `undefined` and so there is an error.
It looks like we need a generic type for `useFetcher`. I tried `useFetcher<typeof action>()`, but that does not work.
What would be the good way to handle this now?

**Comment by kiliman** [comment]
Add `typeof action` to your `useFetcher` call.

```ts
export default function Route() {
  const fetcher = useFetcher<typeof action>();
  // ...
}
```

**Reply by wild-lotus** [reply]
As I mentioned, that did not really work for me. With that, the resulting type of `fetcher.data` is
```
(property) data: JsonifyObject<{
    readonly headers: Headers;
    readonly ok: boolean;
    readonly redirected: boolean;
    readonly status: number;
    readonly statusText: string;
    readonly type: ResponseType;
    readonly url: string;
    readonly body: ReadableStream<...> | null;
    readonly bodyUsed: boolean;
}> | null | undefined

**Reply by sergiodxa** [reply]
Instead of returning `null` try to return `json(null)`

**Reply by wild-lotus** [reply]
Thank you @sergiodxa , that looked promising, but actually as I was about to try it, `useFetcher<typeof action>` did work this time without any other extra changes! 🤷‍♂️🙇‍♂️
